presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java

```diff
@@ -541,6 +543,22 @@ public Boolean visitParameter(Parameter node, Void context)
             return process(parameters.get(node.getPosition()), context);
         }
 
+        public Boolean visitGroupingOperation(GroupingOperation node, Void context)
+        {
+            Optional<Expression> argumentNotInGroupBy = node.getGroupingColumns().stream()
+                    .filter((argument) -> !expressions.contains(argument))
```

Comment:
>  1) not allowing grouping() to reference output columns when in an order by

Yes. When in order by, you shouldn't use output column in `grouping`, e.g:
```
SELECT a FROM (VALUES 1) t(a) GROUP BY a ORDER BY grouping(a)
```
I know this might be inconvenience for users. In future we might think of relaxing this condition for `grouping` so that the example I put here works as expected when output columns are direct aliases of input symbols. For now let's make this strict. Otherwise grouping rewriter won't work (resolved fields won't match). What do you think @martint about relaxing scope rules (pass-through column resolution for `x as x`) for `grouping`?

In general, queries like:
```
SELECT x FROM (VALUES 1) t(x) GROUP BY x ORDER BY sum(x)
```
are illegal since `x` in `sum(x)` references output column which is not aggregated.

Here you can do the validation with those lines:
```
                    // ensure that no output fields are referenced from ORDER BY clause
                    if (orderByScope.isPresent()) {
                        node.getGroupingColumns().stream().forEach(AggregationAnalyzer.this::verifyNoOrderByReferencesToOutputColumns);
                    }
```